### PR TITLE
DocumentationAPI: Support for :raises xyz:

### PR DESCRIPTION
**Pull request by damngamerz**
Add support for :raises xyz:
Add subsequent test cases.
Closes https://github.com/coala/coala/issues/4279

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by damngamerz** [comment]
CC @SanketDG @NiklasMM

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 1]
```diff
@@ -85,12 +86,15 @@ def parse(self):
...
         """
         Parses documentation based on parameter and return symbols.
 
         :param param_identifiers:
             A tuple of two strings with which a parameter starts and ends.
+        :param excp_identifiers:
+            A tuple of two strings with which a exception starts and ends.
         :param return_identifiers:
             The string with which a return description starts.
         :return:
```
Comment (NiklasMM) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: The return description needs updating, since `ExceptionValue` can now also be in the return value.

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 2]
```diff
@@ -121,6 +125,17 @@ def _parse_documentation_with_symbols(self, param_identifiers,
                 param_desc = splitted[1]
                 parsed.append(self.Parameter(name=cur_param, desc=param_desc))
 
+            elif stripped_line.startswith(excp_identifiers[0]):
+                parse_mode = self.ExceptionValue
+                excp_offset = line.find(
+                    excp_identifiers[0]) + len(excp_identifiers[0])
```
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: yup i guess but here we aren't using the `stripped_line`.
So the case may differ for something like:
```
    """
    This is good
        :param xyz: abc
    """
```
see the conditions for `params` im just following the same road :wink:

Comment (NiklasMM) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: oh, yes got it. But why not use the `stripped_line`?

Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Oh yes! that also has a case, what if there's an indentation in metadata's(params,exceps...) desc.
like
```
    """
    This is good

    :param xyz:
        param definition
            1. no idea if someone will write like this.
            2. model point
    """
```
`line` is use to preserve the indentation instead of `stripped_line`.

**Comment by damngamerz** [comment]
@NiklasMM Amended accordingly :+1: . Regarding in-line comments for `if-else` block, the comment to each `else-if` would only create redundant comments, so i have added it to the part wherever i thought it was necessary. The other parts works just the similar way and are easily understood. Ping me if i miss something.

**Comment by NiklasMM** [comment]
ack 42a8755

**Comment by damngamerz** [comment]
@NiklasMM Regarding the DocumentationCommentTest.py I have added 2 tests one is the default one and other is the file test check `default.py`
https://github.com/coala/coala/pull/4281/commits/3a130b42c2b5402061f7b170369febbdc7234312#diff-a2a41bd1ec55c314cf3e795b281775ef

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 3]
```diff
@@ -12,6 +12,7 @@ class DocumentationComment:
     inside source-code, like position etc.
     """
     Parameter = namedtuple('Parameter', 'name, desc')
+    ExceptionValue = namedtuple('ExceptionValue', 'name, desc')
```
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: let's just rename it to exception

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 4]
```diff
@@ -85,18 +86,21 @@ def parse(self):
...
+        :param excep_identifiers:
+            A tuple of two strings with which an exception starts and ends.
         :param return_identifiers:
             The string with which a return description starts.
         :return:
             The list of all the parsed sections of the documentation. Every
-            section is a namedtuple of either ``Description`` or ``Parameter``
-            or ``ReturnValue``.
+            section is a namedtuple of either ``Description``, ``Parameter``,
+            ``ExceptionValue`` or ``ReturnValue``.
```
Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: same

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 3 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: `Exception` is a class in python. It is used as the base class for all exceptions. I guess i cant re-name it to that :stuck_out_tongue:  
https://www.tutorialspoint.com/python/python_exceptions.htm

Comment (SanketDG) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: oh wow totally forgot about that

let's keep it then? do you have a better suggestion?

`ReturnValue` has a value, so it made sense.

If your suggestion also involves changing `Parameter`, then that's fine.

Comment (NiklasMM) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Well, technically you could. But it would create nothing but confusion. :stuck_out_tongue:

File: coalib/bearlib/languages/documentation/default.coalang [hunk 5]
```diff
@@ -2,17 +2,23 @@
...
+exception_start = :raises\ # here's a space
+exception_end = :
 return_sep = :return:
 
 [JAVA]
 doc-marker1 = /**, \ *, \ */
 doc-marker2 = /**, , \ */
 param_start = @param\ \ # here's a space
 param_end = \ # here's a space
+exception_start = @raises\ \ # here's a space
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/default.coalang]: Since you are modifying this, you should also modify test data `default.java` and test case.

File: tests/bearlib/languages/documentation/DocumentationCommentTest.py [hunk 6]
```diff
@@ -114,6 +117,15 @@ def test_params_default(self):
                     self.Parameter(name='test', desc='  test description2 \n')]
         self.check_docstring(doc, expected)
 
+    def test_excps_default(self):
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: write exception properly instead of `excps`.

And why have a 'default' with doc like ` :raises test:  test description1 \n` , which has inappropriate whitespace everywhere.
Why not `    :raises test: test description1\n` which is proper whitespace.

Then add other tests with weird extra whitespace to show how the case behaves when it sees invalid whitespace.

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py [hunk 7]
```diff
@@ -32,3 +32,11 @@ def best_docstring(param1, param2):
     # @param param2 Parameter 2
     # @return Nothing
         return None
+
+def docstring_find(filename):
+    ## This is dummy docstring find function!
+    #
+    # @param filename contains filename.
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py]: IIRC in Doxygen, "contains filename." here should be "Contains filename." (the blank line above also looks wrong , and the `@` should be under the "This", but those will require fixes beyond the scope of this patch)

File: tests/bearlib/languages/documentation/DocumentationCommentTest.py (hunk 6 above)
Comment (damngamerz) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: @jayvdb This is the intended behaviour. Right now extra white spaces are preserved. As this should be later handled by bears(analysing routines). Anyway, removed those inappropriate whitespace and I m amending  a new test case with improper whitespace in `default.py` :+1:

File: coalib/bearlib/languages/documentation/default.coalang (hunk 5 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/default.coalang]: Amended to `default.java`

File: tests/bearlib/languages/documentation/DocumentationCommentTest.py (hunk 6 above)
Comment (damngamerz) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: https://github.com/coala/coala/pull/4281/files#diff-603a58b800b18edbad56a01690a21e05R172
here it is.

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py (hunk 7 above)
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py]: Not done

Comment (damngamerz) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py]: @jayvdb even previous cases are not aligned. should i align all ?

Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py]: Just make sure you code is awesome. And raise newcomer issues for other improvements

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 8]
```diff
@@ -85,18 +86,21 @@ def parse(self):
                 ' has not been implemented yet'.format(self))
 
     def _parse_documentation_with_symbols(self, param_identifiers,
+                                          excep_identifiers,
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Please use snake_case for variables, args, etc, which properly done excludes invented words abbreviating/shortening real words.

See https://www.python.org/dev/peps/pep-0008/#function-and-method-arguments which recommends against "abbreviation or spelling corruption".

Of course it is never that simple.  e.g. `param_` and `doc_` are acceptable, despite being a shortened word. ;-)

A good way to catch yourself, is look if your shortened word has been used elsewhere in the codebase. i.e. `git grep excep_`.  If not, beware, you are attempting to create a new 'custom', possibly inconsistent with the name used elsewhere for the same concept.

But with "exception", I mentioned this in your previous review: https://github.com/coala/coala/pull/4281#discussion_r119154335 , where you used `excps` instead.

This shouldnt be necessary -- each review / re-submit cycle should include a fresh review by you, ensuring your entire patch has fixed the issues identified in previous reviews, especially on the same PR.  Understand and internalise these coding standards.  They are your bible ;-)

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 9]
```diff
@@ -85,18 +86,21 @@ def parse(self):
...
             A tuple of two strings with which a parameter starts and ends.
+        :param excep_identifiers:
+            A tuple of two strings with which an exception starts and ends.
         :param return_identifiers:
             The string with which a return description starts.
         :return:
             The list of all the parsed sections of the documentation. Every
-            section is a namedtuple of either ``Description`` or ``Parameter``
-            or ``ReturnValue``.
+            section is a namedtuple of either ``Description``, ``Parameter``,
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: either words "named tuple" or literal "``namedtuple``"

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 10]
```diff
@@ -113,26 +117,50 @@ def _parse_documentation_with_symbols(self, param_identifiers,
...
                 splitted = line[param_offset:].split(param_identifiers[1], 1)
                 cur_param = splitted[0].strip()
 
                 param_desc = splitted[1]
+                # parsed section is added to the final list.
                 parsed.append(self.Parameter(name=cur_param, desc=param_desc))
 
+            elif stripped_line.startswith(excep_identifiers[0]):
+                parse_mode = self.ExceptionValue
+                excep_offset = line.find(
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: variable names .. per prior comment ... and occurs many times in this patch.

(but `cur_` previous is ok for variable names if you prefer it, but it would be inappropriate for arg names, and unless there is a good reason `current_` should be used.)

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 11]
```diff
@@ -198,6 +226,11 @@ def from_metadata(cls, doccomment, docstyle_definition,
                                   section.name +
                                   docstyle_definition.metadata.param_end)
 
+            elif isinstance(section, cls.ExceptionValue):
+                assembled_doc += (docstyle_definition.metadata.exception_start
+                                  + section.name
```
Comment (jayvdb) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: yay for using `+` after a line break, whereas the previous branch used `+` before the line break.

your code is correct (https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator)

That part of pep 8 was modified about a year ago, iirc.

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py (hunk 7 above)
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py]: Still not done?

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py [hunk 12]
```diff
@@ -37,3 +37,30 @@ def best_docstring(param1, param2):
...
+    :param sideA:
+        length of sideA       
+    :param sideB:
+        length of sideB    
+    :param sideC:
+        length of sideC  
+
+    :return: returns perimeter
+    """
+    return sideA+sideB+sideC
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py]: spaces around `+` per PEP 8.

**Comment by damngamerz** [comment]
Thanks @jayvdb those were good finds and i learned a new thing from that :wink: . All amended :+1: 
Tell me if you find anything else.

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py [hunk 13]
```diff
@@ -37,3 +37,30 @@ def best_docstring(param1, param2):
              Cut to the Next Line.
     """
     return None
+
+def docstring_find(filename):
+    """
+    This is dummy docstring find function!
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py]: You didnt run `pydocstyle` on this, otherwise you would see errors.
(You'll need to ignore errors in pre-existing code; oh how ugly and confusing.)

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py [hunk 14]
```diff
@@ -37,3 +37,30 @@ def best_docstring(param1, param2):
...
+
+    :param filename:
+        contains filename
+    :raises FileNotFoundError:
+        Raised when the given file name was not found.
+
+    :return: returns all possible docstrings in a file.
+    """
+
+def foobar_triangle(sideA, sideB, sideC):
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py]: snake_case arguments

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py [hunk 15]
```diff
@@ -32,3 +32,11 @@ def best_docstring(param1, param2):
     # @param param2 Parameter 2
     # @return Nothing
         return None
+
+def docstring_find(filename):
+    ## This is dummy docstring find function!
+    #
+    #  @param filename contains filename.
+    #  @raises FileNotFoundError raises when filename is not found.
+    #  @return Nothing
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py]: Nothing is Sentence case, but the other two are lower case.

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py [hunk 16]
```diff
@@ -37,3 +37,30 @@ def best_docstring(param1, param2):
              Cut to the Next Line.
     """
     return None
+
+def docstring_find(filename):
+    """
+    This is dummy docstring find function!
+
+    :param filename:
+        contains filename
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py]: In the pre-existing examples in this file , the param description is Sentence case.  Is that the 'default' style?
Your examples are inconsistent.  What should the default style be?

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py [hunk 17]
```diff
@@ -37,3 +37,30 @@ def best_docstring(param1, param2):
...
+    :raises FileNotFoundError:
+        Raised when the given file name was not found.
+
+    :return: returns all possible docstrings in a file.
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py]: Sentence case here too?

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py [hunk 18]
```diff
@@ -37,3 +37,30 @@ def best_docstring(param1, param2):
...
+    """
+
+def foobar_triangle(sideA, sideB, sideC):
+    """
+    This returns perimeter of a triangle!   
+
+    :param sideA:
+        length of sideA
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py]: Sentence case...?

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py (hunk 16 above)
Comment (damngamerz) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py]: Ok so I was correct, only descriptions are Sentence case. Rest all desc. of params, exceptions, returns are lower case. Here check [PEP-257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings). Also same is being followed in [Java docstyle](http://www.oracle.com/technetwork/articles/java/index-137868.html). But still there are some inconsistencies in my implementation fixing them. I will create newcomers issue for the pre-existing examples :+1:

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py [hunk 19]
```diff
@@ -32,3 +32,11 @@ def best_docstring(param1, param2):
     # @param param2 Parameter 2
     # @return Nothing
         return None
+
+def docstring_find(filename):
+    ## This is dummy docstring find function!
```
Comment (jayvdb) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/doxygen.py]: `!`

**Comment by jayvdb** [comment]
ack fa26911 6abf8a0

**Comment by jayvdb** [comment]
@rultor merge
